### PR TITLE
Add new CLI for PG drop counters part of counterpoll utility

### PR DESCRIPTION
**Pull request by ayurkiv-nvda**
Enhance counterpoll utility to support PG drop counters. 
Add new CLI to enable/disable it.

Signed-off-by: Andriy Yurkiv <ayurkiv@nvidia.com>

**depends on:**
https://github.com/Azure/sonic-buildimage/pull/6444
https://github.com/Azure/sonic-swss/pull/1600

**merge priority:**
https://github.com/Azure/sonic-swss/pull/1600
https://github.com/Azure/sonic-utilities/pull/1355
https://github.com/Azure/sonic-buildimage/pull/6444

**- What I did**
Added new option for "counterpoll" utility

**- How I did it**
Enhance counterpoll utility for SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS 
Add new CLI command and extend the show command.

**- How to verify it**
To test it use QoS test. run the pfcxontest and Pfcxofftest, drop can be triggered in either of the test.
It runs in ptf32 topo and requires RPC image.
```
admin@arc-switch1041:~$ counterpoll pg-drop enable  --> to enable the new counter
admin@arc-switch1041:~$ counterpoll show  --> check new INGRESS_PG_STAT_DROP counter status
```
Check counters
```
admin@arc-switch1041:~$ redis-cli -n 2
127.0.0.1:6379[2]> HGETALL COUNTERS:oid:0x1a000000000062
1) "SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES"
2) "0"
3) "SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES"
4) "0"
5) "SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS"
6) "0"

```

**- Previous command output (if the output of a command-line utility has changed)**

```
admin@arc-switch1041:~$ counterpoll show
Type                        Interval (in ms)    Status
--------------------------  ------------------  --------
QUEUE_STAT                  default (10000)     enable
PORT_STAT                   default (1000)      enable
PORT_BUFFER_DROP            default (60000)     enable
RIF_STAT                    default (1000)      enable
QUEUE_WATERMARK_STAT        default (10000)     enable
PG_WATERMARK_STAT           default (10000)     enable
BUFFER_POOL_WATERMARK_STAT  default (10000)     enable
```

```
admin@arc-switch1041:~$ counterpoll
.......
Commands:
  config-db         Config DB counter commands
  port              Queue counter commands
  port-buffer-drop  Port buffer drop counter commands
  queue             Queue counter commands
  rif               RIF counter commands
  show              Show the counter configuration
  watermark         Watermark counter commands
```

**- New command output (if the output of a command-line utility has changed)**
```

admin@arc-switch1041:~$ counterpoll show
Type                        Interval (in ms)    Status
--------------------------  ------------------  --------
QUEUE_STAT                  default (10000)     enable
PORT_STAT                   default (1000)      enable
PORT_BUFFER_DROP            default (60000)     enable
RIF_STAT                    default (1000)      enable
QUEUE_WATERMARK_STAT        default (10000)     enable
PG_WATERMARK_STAT           default (10000)     enable
PG_DROP_STAT                default (10000)     enable         <--- new line
BUFFER_POOL_WATERMARK_STAT  default (10000)     enable

```

```
admin@arc-switch1041:~$ counterpoll
Usage: counterpoll [OPTIONS] COMMAND [ARGS]...
...............
Commands:
  config-db         Config DB counter commands
  pg-drop           Ingress PG drop counter commands     <--- new command
  port              Queue counter commands
  port-buffer-drop  Port buffer drop counter commands
  queue             Queue counter commands
  rif               RIF counter commands
  show              Show the counter configuration
  watermark         Watermark counter commands
```

### CODE REVIEW

**Comment by ayurkiv-nvda** [comment]
retest this please

**Comment by volodymyrsamotiy** [comment]
@neethajohn, could you you please review?

**Comment by liat-grozovik** [comment]
@neethajohn if this is needed for 201911 please add the proper label

**Comment by liat-grozovik** [comment]
@ayurkiv-nvda the PR description is missing some information. I have modified it a little bit but I believe the 'how to test' is should be extended with how to cause the drop counter to be incremented and then check validity.

You are referring to the show command for the counters which is great but should you verify the reset counter command as well?

Please all add content to the section of 'New command output (if the output of a command-line utility has changed)' as you do have a new command.

**Comment by neethajohn** [comment]
To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.

File: tests/counterpoll_test.py [hunk 1]
```diff
@@ -54,6 +55,20 @@ def test_port_buffer_drop_interval_too_short(self):
         assert result.exit_code == 2
         assert expected in result.output
 
+    def test_pg_drop_interval(self):
+        runner = CliRunner()
+        result = runner.invoke(counterpoll.cli.commands["pg-drop"].commands["interval"], ["10000"])
+        print(result.output)
+        assert result.exit_code == 0
```
Comment (neethajohn) [inline tests/counterpoll_test.py:62]: Need to include the check if the interval field is changed to the new value in config db

File: tests/counterpoll_test.py [hunk 2]
```diff
@@ -76,6 +91,13 @@ def test_update_counter_config_db_status(self, status, _get_config_db_file):
             for counter, counter_config in config_db["FLEX_COUNTER_TABLE"].items():
                 assert counter_config["FLEX_COUNTER_STATUS"] == status
 
+    @pytest.mark.parametrize("status", ["disable", "enable"])
+    def test_update_pg_drop_status(self, status):
+        runner = CliRunner()
+        result = runner.invoke(counterpoll.cli.commands["pg-drop"].commands[status])
+        print(result.output)
+        assert result.exit_code == 0
```
Comment (neethajohn) [inline tests/counterpoll_test.py:99]: Please also validate that 'flex_counter_status' is accordingly updated in config db

**Comment by ayurkiv-nvda** [comment]
> To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.

Currently 'counterpoll' utility provide  few command to work with it.
'counterpoll show' - print all counters info in one table. We have general test for it 
'counterpoll <counter command> enable'
'counterpoll <counter command> disable'
'counterpoll <counter command> interval'

There is no possibility to show only specific counter.
The same with clearing. There is no possibility to clean any counters via counterpoll cli

**Comment by ayurkiv-nvda** [comment]
> @ayurkiv-nvda the PR description is missing some information. I have modified it a little bit but I believe the 'how to test' is should be extended with how to cause the drop counter to be incremented and then check validity.
> 
> You are referring to the show command for the counters which is great but should you verify the reset counter command as well?
> 
> Please all add content to the section of 'New command output (if the output of a command-line utility has changed)' as you do have a new command.

added

**Comment by neethajohn** [comment]
> > To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.
> 
> Currently 'counterpoll' utility provide few commands to work with it.
> 'counterpoll show' - print all counters info in one table. We already have general test for it
> 'counterpoll < counter type> enable'
> 'counterpoll < counter type> disable'
> 'counterpoll < counter type> interval' - three option for each counter type
> 
> There is no possibility to show only specific counter.
> The same with clearing. There is no possibility to clean any counters via counterpoll cli

You cannot do it through 'counterpoll'.  You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'

**Comment by ayurkiv-nvda** [comment]
> > > To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.
> > 
> > 
> > Currently 'counterpoll' utility provide few commands to work with it.
> > 'counterpoll show' - print all counters info in one table. We already have general test for it
> > 'counterpoll < counter type> enable'
> > 'counterpoll < counter type> disable'
> > 'counterpoll < counter type> interval' - three option for each counter type
> > There is no possibility to show only specific counter.
> > The same with clearing. There is no possibility to clean any counters via counterpoll cli
> 
> You cannot do it through 'counterpoll'. You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'

Can we come up with CLI  for this counters? This was not considered at the planning stage and will take a time to implement.
What about this command for **show**:
 **show priority-group dropped_packets**
```
        Port    PG0    PG1    PG2    PG3    PG4    PG5    PG6    PG7
-----------  -----  -----  -----  -----  -----  -----  -----  -----
  Ethernet0      0      0      0      0      0      0      0      0
  Ethernet4      0      0      0      0      0      0      0      0
  Ethernet8      0      0      0      0      0      0      0      0
 ...............................................................
Ethernet116      0      0      0      0      0      0      0      0
Ethernet120      0      0      0      0      0      0      0      0
Ethernet124      0      0      0      0      0      0      0      0
```
And this for **clear**:
**sonic-clear priority-group dropped_packet**
 ?

File: tests/counterpoll_test.py (hunk 1 above)
Comment (ayurkiv-nvda) [reply tests/counterpoll_test.py]: Done

File: tests/counterpoll_test.py (hunk 2 above)
Comment (ayurkiv-nvda) [reply tests/counterpoll_test.py:96]: Done

**Comment by neethajohn** [comment]
> > > > To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.
> > > 
> > > 
> > > Currently 'counterpoll' utility provide few commands to work with it.
> > > 'counterpoll show' - print all counters info in one table. We already have general test for it
> > > 'counterpoll < counter type> enable'
> > > 'counterpoll < counter type> disable'
> > > 'counterpoll < counter type> interval' - three option for each counter type
> > > There is no possibility to show only specific counter.
> > > The same with clearing. There is no possibility to clean any counters via counterpoll cli
> > 
> > 
> > You cannot do it through 'counterpoll'. You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'
> 
> Can we come up with CLI for this counters? This was not considered at the planning stage and will take a time to implement.
> What about this command for **show**:
> **show priority-group dropped_packets**
> 
> ```
>         Port    PG0    PG1    PG2    PG3    PG4    PG5    PG6    PG7
> -----------  -----  -----  -----  -----  -----  -----  -----  -----
>   Ethernet0      0      0      0      0      0      0      0      0
>   Ethernet4      0      0      0      0      0      0      0      0
>   Ethernet8      0      0      0      0      0      0      0      0
>  ...............................................................
> Ethernet116      0      0      0      0      0      0      0      0
> Ethernet120      0      0      0      0      0      0      0      0
> Ethernet124      0      0      0      0      0      0      0      0
> ```
> 
> And this for **clear**:
> **sonic-clear priority-group dropped_packet**
> ?

How about
show priority-group drop counters
sonic-clear priority-group drop counters

**Comment by liat-grozovik** [comment]
@Neetha John<mailto:Neetha.John@microsoft.com> may I suggest to differ this req from the PR? As this was not in the initial discussion and there is no specific CLI for each type of counter we should add it as part of new PR and have this done.

**Review by neethajohn — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@ayurkiv-nvda can you please update submodule pointer so we can move forward with the next PR to be merged?

File: counterpoll/main.py [hunk 3]
```diff
@@ -123,6 +124,45 @@ def disable():
     port_info['FLEX_COUNTER_STATUS'] = DISABLE
     configdb.mod_entry("FLEX_COUNTER_TABLE", PORT_BUFFER_DROP, port_info)
 
+# Ingress PG drop packet stat
+@cli.group()
+@click.pass_context
+def pg_drop(ctx):
+    """  Ingress PG drop counter commands """
+    ctx.obj = swsssdk.ConfigDBConnector()
```
Comment (qiluo-msft) [inline counterpoll/main.py:132]: >swsssdk [](start = 14, length = 7)

swsssdk is deprecately largely in sonic-utilities. Could you adapt the new `ConfigDBConnector` as in latest code?

Comment (liat-grozovik) [reply counterpoll/main.py:132]: is it also deprecated on 202012? this PR should go to 202012 as well thus need to know if the change should be done on specific branch only
@ayurkiv-nvda  FYI

**Pull request merged**